File: plugins/OllyLua/read_memory_dialog.lua
local function parse_address(text)
    text = tostring(text or ""):gsub("^%s+", ""):gsub("%s+$", "")
    if text == "" then
        return nil
    end
    text = text:gsub("^0[xX]", "")
    return tonumber(text, 16) or tonumber(text, 10)
end

local function hex_dump(data, base)
    local lines = {}
    for i = 1, #data, 16 do
        local hex = {}
        local ascii = {}
        local n = math.min(16, #data - i + 1)
        for j = 0, n - 1 do
            local b = data:byte(i + j)
            hex[#hex + 1] = string.format("%02X", b)
            ascii[#ascii + 1] = (b >= 32 and b <= 126) and string.char(b) or "."
        end
        lines[#lines + 1] = string.format("%016X  %-47s  %s", base + i - 1, table.concat(hex, " "), table.concat(ascii))
    end
    return table.concat(lines, "\r\n")
end

local address_text = z0dbg.input_text("Read memory", "Address, for example 401000 or 0x401000:", "401000")
if not address_text then
    return
end

local address = parse_address(address_text)
if not address then
    z0dbg.show_text("Read memory", "Invalid address: " .. tostring(address_text))
    return
end

local size_text = z0dbg.input_text("Read memory", "Size in bytes:", "100")
if not size_text then
    return
end

local size = tonumber(size_text) or parse_address(size_text)
if not size or size <= 0 then
    z0dbg.show_text("Read memory", "Invalid size: " .. tostring(size_text))
    return
end

local data, err = z0dbg.read_memory(address, size)
if not data then
    z0dbg.show_text("Read memory", "Read failed: " .. tostring(err))
    return
end

local title = string.format("Memory %016X, %d bytes", address, #data)
z0dbg.show_text(title, hex_dump(data, address))

File: x86/plugins/OllyPython/read_memory_dialog.py
import z0dbg


def parse_address(text):
    text = str(text or "").strip()
    if not text:
        return None
    try:
        return int(text, 0)
    except ValueError:
        try:
            return int(text, 16)
        except ValueError:
            return None


def hex_dump(data, base):
    lines = []
    for offset in range(0, len(data), 16):
        chunk = data[offset:offset + 16]
        hex_part = " ".join("{:02X}".format(b) for b in chunk)
        ascii_part = "".join(chr(b) if 32 <= b <= 126 else "." for b in chunk)
        lines.append("{:016X}  {:<47}  {}".format(base + offset, hex_part, ascii_part))
    return "\r\n".join(lines)


def main():
    address_text = z0dbg.input_text("Read memory", "Address, for example 0x00007FF600001000:", "0x401000")
    if not address_text:
        return

    address = parse_address(address_text)
    if address is None:
        z0dbg.show_text("Read memory", "Invalid address: {}".format(address_text))
        return

    size_text = z0dbg.input_text("Read memory", "Size in bytes:", "100")
    if not size_text:
        return

    size = parse_address(size_text)
    if size is None or size <= 0:
        z0dbg.show_text("Read memory", "Invalid size: {}".format(size_text))
        return

    data = z0dbg.read_memory(address, size)
    if not data:
        z0dbg.show_text("Read memory", "Read failed. Check that the address belongs to the debuggee process.")
        return

    z0dbg.show_text("Memory {:016X}, {} bytes".format(address, len(data)), hex_dump(data, address))


main()
